### ISSUE TITLE
[BUG] `@tensorclass` silently skips `__post_init__` under `torch.compile`, producing wrong output

### ISSUE THREAD

**Issue by peterdsharpe** [pytorch/tensordict#1708]
## Summary

Since v0.12.0, `__post_init__` is not called when a `@tensorclass` is constructed inside a `torch.compile`-traced function. There is no error: the tensorclass is returned with fields exactly as they were passed to the constructor, and any normalization, derivation, or validation the user wrote in `__post_init__` is silently dropped. Compiled output diverges from eager output.

## Minimal reproducer

```python
import torch
from tensordict import tensorclass


@tensorclass
class Foo:
    x: torch.Tensor

    def __post_init__(self):
        self.x = self.x * 2


print("Eager   :", Foo(x=torch.ones(3)).x)                              # tensor([2., 2., 2.])
print("Compiled:", torch.compile(lambda x: Foo(x=x).x)(torch.ones(3)))  # tensor([1., 1., 1.])
```

Output:

```
Eager   : tensor([2., 2., 2.])
Compiled: tensor([1., 1., 1.])
```

## Root cause

PR [#1552](https://github.com/pytorch/tensordict/pull/1552) replaced the unconditional `__init__(self, **kwargs)` call in `_init_wrapper` with a fast-path bypass branch. The `__post_init__` invocation at the end of that branch is guarded by `not is_compiling()`:

```python
# tensordict/tensorclass.py @ v0.12.3, line 1412
if not is_compiling() and hasattr(type(self), "__post_init__"):
    self.__post_init__()
```

Under `torch.compile`, the guard short-circuits and `__post_init__` is never called. The symmetric skip exists on the `_from_tensordict` path at [line 1664](https://github.com/pytorch/tensordict/blob/v0.12.3/tensordict/tensorclass.py#L1664).

Pre-#1552 (v0.11.0 and earlier), the wrapper unconditionally called the user `__init__(self, **kwargs)` ([v0.11.0 line 1295](https://github.com/pytorch/tensordict/blob/v0.11.0/tensordict/tensorclass.py#L1295)), which in turn called `__post_init__` as part of dataclass semantics. The skip is purely a side effect of bypassing the user `__init__`.

## Version timeline

| tensordict | Compiled output of repro | Notes |
|------------|--------------------------|-------|
| 0.10.0 | `tensor([2., 2., 2.])` (correct) | unconditional `__init__(self, **kwargs)` |
| 0.11.0 | `tensor([2., 2., 2.])` (correct) | same |
| 0.12.0 | `tensor([1., 1., 1.])` (**wrong**) | PR [#1552](https://github.com/pytorch/tensordict/pull/1552) |
| 0.12.1, 0.12.2, 0.12.3 | wrong | code path unchanged |
| `main` @ HEAD | wrong | same guard at [tensorclass.py:1418](https://github.com/pytorch/tensordict/blob/main/tensordict/tensorclass.py) |

## Expected behavior

`__post_init__` should be called under `torch.compile`, matching eager semantics. The dataclass contract is that `__post_init__` always runs as part of construction. Users cannot detect or guard against it being skipped at the call site, because the construction itself doesn't raise — the tensorclass just has the wrong state.

The performance motivation of PR #1552 (avoiding dataclass `__init__` / `__setattr__` dispatch) is unaffected by calling `__post_init__`: it runs *after* the fast direct-dict assignment and adds at most one Python function call per construction.

## Suggested fix

Drop the `not is_compiling()` half of both guards:

```diff
- if not is_compiling() and hasattr(type(self), "__post_init__"):
+ if hasattr(type(self), "__post_init__"):
      self.__post_init__()
```

at [`tensorclass.py:1412`](https://github.com/pytorch/tensordict/blob/v0.12.3/tensordict/tensorclass.py#L1412) (bypass branch in `_init_wrapper`) and the symmetric guard at [`tensorclass.py:1664`](https://github.com/pytorch/tensordict/blob/v0.12.3/tensordict/tensorclass.py#L1664) (`_from_tensordict`).

If `__post_init__` was intentionally guarded because Dynamo has trouble tracing it, the alternative is to take the `_has_custom_setattr=True` slow path (call user `__init__` -> calls `__post_init__`) whenever `is_compiling()` is true. The slow path was the only path pre-#1552 and worked correctly under compile.

## System info

```
Python      : 3.12.12
torch       : 2.11.0+cu130
tensordict  : 0.12.3
OS          : Linux x86_64
```

## Checklist

- [x] I have checked that there is no similar issue in the repo (searched `post_init`, `__post_init__`, `tensor_only compile`; the only `__post_init__` issue is [#172](https://github.com/pytorch/tensordict/issues/172), the original feature request from 2023)
- [x] I have read the [documentation](https://github.com/pytorch/tensordict/tree/main/docs/)
- [x] I have provided a minimal working example to reproduce the bug
